src/Factory/RepositoryContainer.php

```diff
@@ -0,0 +1,84 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Yiisoft\Yii\Cycle\Factory;
+
+use Cycle\ORM\ORMInterface;
+use Cycle\ORM\RepositoryInterface;
+use Cycle\ORM\SchemaInterface;
+use Psr\Container\ContainerInterface;
+use Yiisoft\Definitions\Exception\NotFoundException;
+use Yiisoft\Definitions\Exception\NotInstantiableClassException;
+
+use function is_string;
+
+final class RepositoryContainer implements ContainerInterface
+{
+    private ContainerInterface $rootContainer;
+    private ORMInterface $orm;
+
+    private bool $rolesBuilt = false;
+    private array $roles = [];
+
+    private array $instances;
+
+    public function __construct(ContainerInterface $rootContainer)
+    {
+        $this->rootContainer = $rootContainer;
+    }
+
+    public function get($id)
+    {
+        if (isset($this->instances[$id])) {
+            return $this->instances[$id];
+        }
+
+        if ($this->has($id)) {
+            return $this->instances[$id] = $this->makeRepository($this->roles[$id]);
+        }
+
+        throw new NotFoundException($id);
```

Comment:
The same https://github.com/yiisoft/yii-cycle/pull/102#pullrequestreview-727209166

> It is preferable to throw several friendly errors here:
>
> - "ambiguous repository definition" when the repository is used by several roles;
> - "the repository is not defined in the ORM Schema"
> - "the requested object is not a repository (doesn't implements RepositoryInterface)"

Reply:
Since it's used within composite container, `get()` that would error will never be called.